HPC Episode

multimeric opened the pull request:
* Added an optional HPC episode, which uses Slurm as an example
* ~~This has been written using the current plan writing system. Once/if #19 is merged, some of this will need re-writing to remove redundancy~~

multimeric committed:
future -> crew

multimeric committed:
Fix typo

multimeric committed:
HPC draft

multimeric committed:
Start worker section

multimeric committed:
Finished draft of heterogenous workers

multimeric committed:
Update dependencies to have crew, finish first draft of HPC

multimeric committed:
Update knit_exit comment

multimeric committed:
Merge branch 'main' of github.com:carpentries-incubator/targets-workshop into hpc

Conflicts:
	renv/profiles/lesson-requirements/renv.lock

jIskCoder commented on episodes/hpc.Rmd:
```diff
@@ -0,0 +1,295 @@
+
+If your analysis involves computationally intensive or long-running tasks such as training machine learning models or processing very large amounts of data, it will quickly become infeasible to use a single machine to run this.
+If you are part of an organisation with access to a High Performance Computing (HPC) cluster, you can easily leverage the numerous machines with Targets to scale up your analysis.
+This differs from the exucution we have learned so far, which spawns extra R processes on the *same machine* to speed up execution.
```
'exucution' should be 'execution'

jIskCoder commented on episodes/hpc.Rmd:
```diff
@@ -0,0 +1,295 @@
+tar_option_set(
+  controller = crew_controller_slurm(
+    workers = 3,
+    script_lines = "module load R"
```
Add an example to show how to add SBATCH options too.

edoyango commented on episodes/hpc.Rmd:
```diff
@@ -0,0 +1,295 @@
+## Advantages of HPC
+
+If your analysis involves computationally intensive or long-running tasks such as training machine learning models or processing very large amounts of data, it will quickly become infeasible to use a single machine to run this.
+If you are part of an organisation with access to a High Performance Computing (HPC) cluster, you can easily leverage the numerous machines with Targets to scale up your analysis.
```
```suggestion
If you have access to a High Performance Computing (HPC) cluster, you can leverage the numerous machines with Targets to scale up your analysis.
```
I'm being pedantic here, but one does not have to be part of an organisation with an HPC facility to be able to use one.

Furthermore, try not to use terms like "easily", "only", "just" and similar words, as required by carpentry instructors + courses. These kind of words downplay the difficulty or complexity of a task, which can be discouraging or off-putting to learners who find the task challenging.

multimeric replied:
Thanks, that's good advice.

edoyango commented on episodes/hpc.Rmd:
```diff
@@ -0,0 +1,295 @@
+## Configuring Targets for Slurm
+
+Fortunately, using HPC is as simple as changing the Targets `controller`.
+In this section we will assume that our HPC uses Slurm as its job scheduler, but you can easily use other schedulers such as PBS/TORQUE, Sun Grid Engine (SGE) or LSF.
```
```suggestion
In this section we will assume that our HPC uses Slurm as its job scheduler, but you can use other schedulers such as PBS/TORQUE, Sun Grid Engine (SGE) or LSF.
```
Could you also elaborate on how this would be done?

edoyango commented on episodes/hpc.Rmd:
````diff
@@ -0,0 +1,295 @@
+  controller = crew_controller_local(workers = 2)
+)
+```
+To configure this for Slurm, we just swap out the controller with a new one from the `crew.cluster` package:
````
```suggestion
To configure this for Slurm, we swap out the controller with a new one from the `crew.cluster` package:
```

edoyango commented on episodes/hpc.Rmd:
```diff
@@ -0,0 +1,295 @@
+Here we are only using two:
+
+  * `workers` sets the number of jobs that are submitted to Slurm to process targets.
+  * `script_lines` adds some lines to the Slurm submit script used by Targets. This is useful for loading Environment Modules and adding `#SBATCH` options.
```
I would suggest first moving `This is useful for loading Environment Modules and adding `#SBATCH` options.` to somewhere after your first demonstration with the `module load R` command. Carpentries prefer that you teach one thing at a time. So, you first introduce the script_lines concept, and then connect it to the Slurm options afterwards.

edoyango commented on episodes/hpc.Rmd:
```diff
@@ -0,0 +1,295 @@
+::: challenge
+## Increasing Resources
+
+Q: How would you modify your `_targets.R` if your targets needed 200GB of RAM?
```
Is there a reason why you chose 200GB? Smaller HPC facilities might not have nodes with this much RAM Choosing a smaller number would help make this more accessible.

multimeric replied:
No particular reason, but this example isn't actually run so I don't think it matters?

edoyango replied:
I guess it isn't critical, since it's just a test of knowledge. But people may want to run it to confirm for themselves, so lowering the value would make it easier.

edoyango commented on episodes/hpc.Rmd:
```diff
@@ -0,0 +1,295 @@
+
+## HPC Workers
+
+Despite what you might expect, `crew` does not submit one Slurm job for each target. 
```
```suggestion
`crew` uses a persistent worker strategy. This means that `crew` does not submit one Slurm job for each target. 
```
Better not to assume what expectations the learner will have.

edoyango commented on episodes/hpc.Rmd:
```diff
@@ -0,0 +1,295 @@
+In our example above we used 3 workers.
+For each worker, `crew` submits a single Slurm job, and these workers will process multiple targets over their lifetime.
+
+We can verify that this has happened using `sacct`:
```
To keep the material as agnostic to the scheduler as possible, I suggest removing this as learners not using Slurm would have no idea what `sacct` does and what the output means.

multimeric replied:
The episode is already reasonably tied to Slurm by virtue of the controller class being Slurm-specific. Considering that, is there much benefit of removing this?

edoyango replied:
```suggestion
We can verify that this has happened using `sacct`, which we can use to query information about our past jobs:
```

Yeah you're probably right that it is not that meaningful to remove it. But perhaps add a bit of an explainer to what the command is doing.

A semi-related question: Are there any monitoring tools available for the targets? I imagine that would also be useful for parallel execution.

multimeric replied:
There are, but they're mostly covered already in the Lifecycle episode, or are extensions of the same functions: https://docs.ropensci.org/targets/reference/index.html#progress

edoyango commented on episodes/hpc.Rmd:
````diff
@@ -0,0 +1,295 @@
+sacct
+```
+
+The upside of this approach is that we don't have to work out the minutae of how long each target takes to build, or what resources it needs.
````
```suggestion
The upside of this approach is that we don't have to know how long each target takes to build, or what resources it needs.
```

edoyango commented on episodes/hpc.Rmd:
````diff
@@ -0,0 +1,295 @@
+```
+
+The upside of this approach is that we don't have to work out the minutae of how long each target takes to build, or what resources it needs.
+It also means that we don't submit a lot of jobs, making our Slurm usage more efficient and easy to monitor.
````
Submitting lots of jobs is not really a problem for an HPC scheduler, but people generally care more about resource utilization. One could argue that the persistent worker strategy is potentially less efficient. I would suggest removing this line.

multimeric replied:
Would it be reasonable to remove the point about efficiency but keep the rest?

edoyango commented on episodes/hpc.Rmd:
```diff
@@ -0,0 +1,295 @@
+The upside of this approach is that we don't have to work out the minutae of how long each target takes to build, or what resources it needs.
+It also means that we don't submit a lot of jobs, making our Slurm usage more efficient and easy to monitor.
+
+The downside of this mechanism is that **the resources of the worker have to be sufficient to build each of your targets**.
```
maybe you mean "the worker have to be sufficient to build all of your targets"? Maybe an example would be more helpful.

multimeric replied:
I used to have an example here, but I converted the example into the following challenge block. Is this not a good approach?

edoyango replied:
I just didn't understand that sentence that well. The example in the challenge is clear, but if I were in the learner's shoes, I would probably be "guessing" the answer to the challenge with the explanation alone.

edoyango replied:
hence why changing "each target" to "all targets" might make more sense, as it expresses that all the targets need to be able to run within the crew's resources.

edoyango commented on episodes/hpc.Rmd:
```diff
@@ -0,0 +1,295 @@
+
+## Heterogeneous Workers
+
+In some cases we may prefer heterogeneous workers, especially if some of our targets need a GPU and others need a CPU.
```
Can you clarify what "heterogeneous" means here?

edoyango commented on episodes/hpc.Rmd:
```diff
@@ -0,0 +1,295 @@
+
+## Configuring Targets for Slurm
+
+Fortunately, using HPC is as simple as changing the Targets `controller`.
```
```suggestion
To adapt Targets to use the Slurm HPC scheduler, we change the `controller`.
```
See comment above about language that can understate effort.

edoyango commented on episodes/hpc.Rmd:
````diff
@@ -0,0 +1,295 @@
+)
+```
+
+There are a number of options you can pass to `crew_controller_slurm()` to fine-tune the Slurm execution, [which you can find here](https://wlandau.github.io/crew.cluster/reference/crew_controller_slurm.html).
````
Could you elaborate on the type of options that are available? E.g., when might you have to use the "script_lines" vs the targets options to control Slurm execution config?
Are there equivalents for other schedulers?

multimeric committed:
Better wording when discussing HPC

Co-authored-by: Edward Yang <94523015+edoyango@users.noreply.github.com>

multimeric committed:
Don't assume learner expectations

Co-authored-by: Edward Yang <94523015+edoyango@users.noreply.github.com>

multimeric committed:
Update episodes/hpc.Rmd

Co-authored-by: Edward Yang <94523015+edoyango@users.noreply.github.com>

multimeric committed:
Update episodes/hpc.Rmd

Co-authored-by: Edward Yang <94523015+edoyango@users.noreply.github.com>

multimeric committed:
Update episodes/hpc.Rmd

Co-authored-by: Edward Yang <94523015+edoyango@users.noreply.github.com>

multimeric committed:
Update episodes/hpc.Rmd

Co-authored-by: Edward Yang <94523015+edoyango@users.noreply.github.com>

multimeric committed:
Code review, and update dependencies

multimeric committed:
Merge branch 'hpc' of github.com:multimeric/targets-workshop into hpc

Conflicts:
	episodes/hpc.Rmd

multimeric committed:
Minor improvements and rewording

edoyango reviewed (commented):
A couple comments after having gone through the preceding episodes.

multimeric committed:
Install crew.cluster

Co-authored-by: Edward Yang <94523015+edoyango@users.noreply.github.com>

multimeric committed:
Elaborate on `sacct`

Co-authored-by: Edward Yang <94523015+edoyango@users.noreply.github.com>

multimeric committed:
Tweaks to the HPC episode

multimeric committed:
Use crew_controller_group, and add memory example

edoyango commented on episodes/hpc.Rmd:
```diff
@@ -254,33 +254,110 @@ crew_controller_slurm(
+tar_plan(
+  tar_target(
+    name = big_memory_task,
+    command = Sys.getenv("SBATCH_MEM_PER_CPU"),
```
```suggestion
    command = Sys.getenv("SLURM_MEM_PER_CPU"),
```

edoyango commented on episodes/hpc.Rmd:
```diff
@@ -254,33 +254,110 @@ crew_controller_slurm(
+  ),
+  tar_target(
+    name = small_memory_task,
+    command = Sys.getenv("SBATCH_MEM_PER_CPU"),
```
```suggestion
    command = Sys.getenv("SLURM_MEM_PER_CPU"),
```

multimeric committed:
Update episodes/hpc.Rmd

Co-authored-by: Edward Yang <94523015+edoyango@users.noreply.github.com>

multimeric committed:
Update episodes/hpc.Rmd

Co-authored-by: Edward Yang <94523015+edoyango@users.noreply.github.com>

joelnitta commented:
@multimeric thanks for your patience with #19, which has now been merged.

Is this PR basically ready to go in terms of content? There was a fair bit of discussion about it earlier, but seems to have quieted down recently (I personally don't use an HPC server so don't have much to comment on).

Also, in the OP you mentioned `Once/if https://github.com/carpentries-incubator/targets-workshop/pull/19 is merged, some of this will need re-writing to remove redundancy`. Does that still need to happen?

multimeric commented:
Yep, the content is all sorted, it's what I presented in July. But indeed I do need to rewrite some of this now that #19 has been merged.

multimeric committed:
Merge branch 'main' of github.com:carpentries-incubator/targets-workshop into hpc

Conflicts:
	episodes/files/functions.R
	renv/profiles/lesson-requirements/renv.lock

multimeric committed:
Merge branch 'hpc' of github.com:multimeric/targets-workshop into hpc

Conflicts:
	episodes/hpc.Rmd

multimeric committed:
Remove some superfluous files

multimeric committed:
Remove unused figures

multimeric committed:
I guess these files should be tracked?

multimeric commented:
Hi @joelnitta, I've merged in the main branch changes now. However I just realised that, since the sandpaper builds are mostly going to be happening on the CI server, it won't be able to use Slurm. Should I:
1. Don't build the HPC lesson on the CI
2. Include pre-computed results, and make all chunks `eval: false`
3. Add slurm to the CI server, e.g. https://github.com/koesterlab/setup-slurm-action

joelnitta commented:
Thanks @multimeric.

>     1. Don't build the HPC lesson on the CI
> 
>     2. Include pre-computed results, and make all chunks `eval: false`
> 
>     3. Add slurm to the CI server, e.g. https://github.com/koesterlab/setup-slurm-action

My preference is `3` > `2` > `1`. The lesson webpage is rendered with CI, so `1` would mean we don't have the HPC lesson on the webpage either (it would only be available when locally rendered).

joelnitta commented on renv/profiles/lesson-requirements/renv/.gitignore:
Why is this deleted? When I try to render this branch, renv is scanning everything in `renv/library` since it is no longer ignored.

joelnitta commented on renv/profiles/lesson-requirements/renv/settings.json:
Same question as above. I think renv settings should be preserved.

joelnitta commented on episodes/hpc.Rmd:
```diff
@@ -0,0 +1,324 @@
+
+If your analysis involves computationally intensive or long-running tasks such as training machine learning models or processing very large amounts of data, it will quickly become infeasible to use a single machine to run this.
+If you have access to a High Performance Computing (HPC) cluster, you can leverage the numerous machines with Targets to scale up your analysis.
+This differs from the exucution we have learned so far, which spawns extra R processes on the *same machine* to speed up execution.
```
Typo: `exucution` -> `execution`

multimeric committed:
Add missing activate file

multimeric committed:
Add missing .gitignore

multimeric committed:
Fix execution typo

multimeric committed:
Start slurm for build

joelnitta committed:
Force dep on {htmlwidgets}

needed for {visNetwork}

joelnitta committed:
Update renv and packages

joelnitta committed:
Update renv.lock

joelnitta commented:
@multimeric 

I realized that I needed to update `renv.lock` in `main`, so I did that and cherry-picked the commits here ([4eb5f06](https://github.com/carpentries-incubator/targets-workshop/pull/18/commits/4eb5f063e3bfdf2b34dc5f807a01e5fd7b9b46be), [9489bea](https://github.com/carpentries-incubator/targets-workshop/pull/18/commits/9489bea236965744de194bc5768a79d2483ac5f6)), as well as make one more update to `renv.lock` (I think due to using CRAN instead of the carpentries r-universe, [e012aaf](https://github.com/carpentries-incubator/targets-workshop/pull/18/commits/e012aafef4de7620a750e93606a0f930094ad766)).

Can you check if you can still build the HPC lesson locally? I can't because I don't have access to a computer with slurm.

multimeric commented:
I can do that, but can we also test if it's working on the CI? I'm not sure if it actually runs the lesson build for pull requests at the moment?

joelnitta commented:
Sure - I assume you mean implementing suggestion `3`?

> Add slurm to the CI server, e.g. https://github.com/koesterlab/setup-slurm-action

multimeric commented:
Oh, I already added that to `.github/workflows/sandpaper-main.yaml`, but I'm not sure if it will even run that workflow for a pull request. Also it doesn't like the fact that I've changed the content and the workflow in the same PR.

joelnitta commented:
How about this: I just set up another repo and merged `hpc` into main over there, let's see how it goes.

GH action is running now... https://github.com/joelnitta/targets-hpc/actions/runs/9867906104

multimeric committed:
Reduce memory requirement

multimeric commented:
No, I think it is a fault in my PR. It's failing at the second workflow which uses different memory requirements, which is likely the issue. I've reduced the memory requirements, but I now realise that it's guaranteed to fail at the GPU workflow since the mini cluster isn't going to have a GPU node.

joelnitta commented:
ah... so perhaps CI is not an option then?

multimeric committed:
Merge branch 'hpc' of github.com:multimeric/targets-workshop into hpc

multimeric commented:
Well not for the last workflow, but that's inside a challenge solution so even if I just paste in the pre-computed result there I don't think that's too bad.

joelnitta commented:
Sounds good, let's try that then. If you push the change here, I can try it again over on my other fork.

multimeric committed:
Re-run CI

multimeric committed:
Merge branch 'main' of github.com:carpentries-incubator/targets-workshop into hpc

multimeric committed:
Revert .github and renv changes

multimeric committed:
Merge branch 'main' of github.com:carpentries-incubator/targets-workshop into hpc

joelnitta commented:
@multimeric 

[markdown build](https://github.com/carpentries-incubator/targets-workshop/actions/runs/12310872381/job/34360132250?pr=18) seems to be hanging

![image](https://github.com/user-attachments/assets/50e740d3-6047-48b4-961c-593e136ec19e)

multimeric commented:
Yeah, feel free to cancel it. I can't.

multimeric commented:
I believe this failed because the lesson involves submitting a job that requires 2 GB of RAM, but the `koesterlab/setup-slurm-action` only provides a single node with 2000 MB of RAM and so it hangs forever while waiting for sufficient memory to become available. I'll fork the action to allow increased memory.

multimeric commented:
Actually can you give me the permission to cancel builds so that I can play around with this?

multimeric commented:
Okay, putting this on hold until https://github.com/koesterlab/setup-slurm-action/pull/7 is merged. Technically we could use my fork of the action, but there's no hurry.

multimeric committed:
Remove the crew installation

multimeric committed:
Don't evaluate the crew_controller_local

multimeric committed:
Use the new slurm_memory_gigabytes_required instead of slurm_memory_gigabytes_per_cpu

multimeric committed:
Merge branch 'main' of github.com:carpentries-incubator/targets-workshop into hpc

multimeric committed:
Use new crew_options_slurm

multimeric committed:
Merge branch 'hpc' of github.com:multimeric/targets-workshop into hpc

multimeric committed:
Fix the lines we show to the user; use SLURM_MEM_PER_NODE when reading
memory; load crew.cluster
